src/ServiceControl.Monitoring/Timings/TimingsStore.cs

```diff
@@ -3,122 +3,162 @@
     using System;
     using System.Collections.Concurrent;
     using System.Collections.Generic;
+    using System.Threading;
     using Metrics.Raw;
 
     public abstract class TimingsStore
     {
-        ConcurrentDictionary<EndpointInstanceId, ConcurrentDictionary<DateTime, MeasurementInterval>> timings = 
-            new ConcurrentDictionary<EndpointInstanceId, ConcurrentDictionary<DateTime, MeasurementInterval>>();
+        ConcurrentDictionary<EndpointInstanceId, Measurement> timings = 
+            new ConcurrentDictionary<EndpointInstanceId, Measurement>();
 
         public void Store(EndpointInstanceId instanceId, LongValueOccurrences message, DateTime now)
         {
-            var instanceData = timings.GetOrAdd(instanceId, _ => new ConcurrentDictionary<DateTime, MeasurementInterval>());
+            var measurement = timings.GetOrAdd(instanceId, _ => new Measurement());
+            measurement.Report(message);
+        }
 
-            for (var i = 0; i < message.Ticks.Length; i++)
+        public EndpointInstanceTimings[] GetTimings(DateTime now)
+        {
+            var result = new List<EndpointInstanceTimings>();
+
+            foreach (var timing in timings)
             {
-                var date = new DateTime(message.BaseTicks + message.Ticks[i], DateTimeKind.Utc);
-                var intervalId = date.RoundDownToNearest(IntervalSize);
+                var instanceId = timing.Key;
+                var measurement = timing.Value;
 
-                instanceData.AddOrUpdate(
-                    intervalId,
-                    _ => new MeasurementInterval(1, message.Values[i]),
-                    (_, b) => b.Update(message.Values[i]));
-            }
+                long totalDuration;
+                long totalMeasurements;
+                var timingIntervals = new TimingInterval[NumberOfHistoricalIntervals];
 
-            if (instanceData.Count > 2 * NumberOfHistoricalIntervals)
-            {
-                var historySize = TimeSpan.FromTicks(IntervalSize.Ticks * NumberOfHistoricalIntervals);
-                var oldestValidInterval = now.Subtract(historySize).RoundDownToNearest(IntervalSize);
+                measurement.ReportTimeIntervals(now, timingIntervals, out totalDuration, out totalMeasurements);
```

Comment:
Would it make sense to create an instance of `EndpointInstanceTimings` and pass it in to be filled in (with totals)?